ddspipe_core/src/cpp/core/DdsPipe.cpp

```diff
@@ -226,62 +288,86 @@ void DdsPipe::discovered_endpoint_(
     {
         if (!RpcTopic::is_service_topic(endpoint.topic))
         {
-            discovered_topic_nts_(utils::Heritable<DdsTopic>::make_heritable(endpoint.topic));
+            discovered_topic_nts_(utils::Heritable<DdsTopic>::make_heritable(
+                        endpoint.topic), endpoint.discoverer_participant_id);
         }
         else if (endpoint.is_server_endpoint())
         {
             // Service server discovered
-            discovered_service_nts_(types::RpcTopic(
+            discovered_service_nts_(RpcTopic(
                         endpoint.topic), endpoint.discoverer_participant_id, endpoint.guid.guid_prefix());
         }
     }
 }
 
-void DdsPipe::removed_endpoint_(
+void DdsPipe::removed_endpoint_nts_(
         const Endpoint& endpoint) noexcept
 {
-    std::lock_guard<std::mutex> lock(mutex_);
-
     logDebug(DDSPIPE, "Endpoint removed/dropped: " << endpoint << ".");
 
-    const DdsTopic& topic = endpoint.topic;
-    if (RpcTopic::is_service_topic(topic))
+    const auto& topic = utils::Heritable<DdsTopic>::make_heritable(endpoint.topic);
+
+    if (!RpcTopic::is_service_topic(endpoint.topic))
     {
-        if (endpoint.is_server_endpoint())
+        if (endpoint.is_writer())
         {
-            // Service server removed/dropped
-            removed_service_nts_(types::RpcTopic(topic), endpoint.discoverer_participant_id,
-                    endpoint.guid.guid_prefix());
+            // An external writer has been removed.
+            // The bridge and the tracks remain intact.
+            return;
         }
+
+        // Remove the subscriber from the topic.
+        auto it_bridge = bridges_.find(topic);
+
+        if (it_bridge == bridges_.end())
+        {
+            // The bridge does not exist. We cannot remove the writer. Exit.
+            return;
+        }
+        else if (remove_unused_entities_)
+        {
+            it_bridge->second->remove_from_tracks(endpoint.discoverer_participant_id);
+        }
+    }
+    else if (endpoint.is_server_endpoint())
+    {
+        // Service server removed/dropped
+        removed_service_nts_(RpcTopic(endpoint.topic), endpoint.discoverer_participant_id, endpoint.guid.guid_prefix());
     }
 }
 
 void DdsPipe::init_bridges_nts_(
-        const std::set<utils::Heritable<types::DistributedTopic>>& builtin_topics)
+        const std::set<utils::Heritable<DistributedTopic>>& builtin_topics)
 {
     for (const auto& topic : builtin_topics)
     {
-        discovered_topic_nts_(topic);
+        discovered_topic_nts_(topic, "");
```

Comment:
The problem is that some tests use an empty discoverer_participant_id tag. I think now, since everything is being taken care of at the ITopic, it makes sense to leave it blank by default and have it modified when there's a discoverer.

Reply:
Yes I understand that, but I believe it would be good to have something different from an empty string. This way we could differentiate (debugging for example) when this value has not been set (default value) or has been lost due to a bug.